### PR TITLE
Another solution to custom getattr/delattr

### PR DESCRIPTION
**Pull request by dmontagu**
I think this is perhaps a better solution to #5946, as it ensures that you can call `super().__getattr__` and `super().__delattr__` when there are private attributes or allowed extra attributes, and things just work.

I think this would be easier to properly document than #6116, and is probably a better user experience either way. But it's a bit quirky, adding a custom class to the bases like this.

Selected Reviewer: @lig

### CODE REVIEW

**Comment by dmontagu** [comment]
Okay, I tried _always_ defining the method, and it led to issues interacting with descriptors. I think probably best not to try to always define it, as I suspect it will be hard to get everything working correctly, and this seems to be enough of an edge case that I think either this PR or #6116 should be preferred since they will likely require less effort to deploy.

File: pydantic/_internal/_model_construction.py [hunk 1]
```diff
@@ -471,7 +470,28 @@ def generate_model_signature(
...
+    """Returns a base class defining a custom __getattr__ and/or __delattr__ for models needing them for the sake
+    of interaction with private attributes and/or `model_config['extra'] == 'allow'`.
+    """
+    extra_base_namespace = {}
+    if config_wrapper.extra == 'allow' or namespace['__private_attributes__']:
+        if not any('__getattr__' in base.__dict__ for base in bases):
+            extra_base_namespace['__getattr__'] = _extra_private_getattr
+    if namespace['__private_attributes__']:
+        if not any('__delattr__' in base.__dict__ for base in bases):
+            extra_base_namespace['__delattr__'] = _extra_private_getattr
```
Comment (lig) [inline pydantic/_internal/_model_construction.py:485]: `getattr` as `delattr`?

**Review by lig — DISMISSED** [review]

File: pydantic/_internal/_model_construction.py [hunk 2]
```diff
@@ -471,7 +470,28 @@ def generate_model_signature(
...
+    bases: tuple[type[Any], ...], config_wrapper: ConfigWrapper, namespace: dict[str, Any]
+) -> type[Any] | None:
+    """Returns a base class defining a custom __getattr__ and/or __delattr__ for models needing them for the sake
+    of interaction with private attributes and/or `model_config['extra'] == 'allow'`.
+    """
+    extra_base_namespace = {}
+    if config_wrapper.extra == 'allow' or namespace['__private_attributes__']:
+        if not any('__getattr__' in base.__dict__ for base in bases):
+            extra_base_namespace['__getattr__'] = _extra_private_getattr
+    if namespace['__private_attributes__']:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/_internal/_model_construction.py:483]: makes sense

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/_internal/_model_construction.py:483]: agreed.

**Review by samuelcolvin — APPROVED** [review]

**Comment by dmontagu** [comment]
please review

**Review by lig — APPROVED** [review]
👍🏻

**Comment by dmontagu** [comment]
Okay, I modified this to just define `__getattr__` and `__delattr__` on BaseModel, which I think is much simpler and easier to understand.

To try to get a sense of how much this might impact actual usage, I ran the full test suite via `pytest -s` with the following patch applied:
```diff
diff --git a/pydantic/main.py b/pydantic/main.py
index 03a87199..f9971f4d 100644
--- a/pydantic/main.py
+++ b/pydantic/main.py
@@ -4,6 +4,7 @@ from __future__ import annotations as _annotations
 import types
 import typing
 import warnings
+from collections import Counter
 from copy import copy, deepcopy
 from typing import Any
 
@@ -58,6 +59,9 @@ __all__ = 'BaseModel', 'create_model'
 _object_setattr = _model_construction.object_setattr
 _Undefined = _fields.Undefined
 
+getattr_count = 0
+getattr_counts = Counter()
+
 
 class BaseModel(metaclass=_model_construction.ModelMetaclass):
     """A base model class for creating Pydantic models.
@@ -317,6 +321,7 @@ class BaseModel(metaclass=_model_construction.ModelMetaclass):
     if not typing.TYPE_CHECKING:  # otherwise mypy will allow arbitrary attribute access
 
         def __getattr__(self, item: str) -> Any:
+            global getattr_count, getattr_counts
             private_attributes = object.__getattribute__(self, '__private_attributes__')
             if item in private_attributes:
                 attribute = private_attributes[item]
@@ -336,6 +341,9 @@ class BaseModel(metaclass=_model_construction.ModelMetaclass):
                     except KeyError as exc:
                         raise AttributeError(f'{type(self).__name__!r} object has no attribute {item!r}') from exc
                 else:
+                    getattr_count += 1
+                    getattr_counts[item] += 1
+                    print(item, getattr_count, getattr_counts)
                     raise AttributeError(f'{type(self).__name__!r} object has no attribute {item!r}')
 
     def __setattr__(self, name: str, value: Any) -> None:
```

in the end, after running close to 4000 tests there were 55 new calls to `BaseModel.__getattr__`, with the following counts for specific values of `item`:
```
{
  "keys": 23,
  "__name__": 6,
  "b": 4,
  "__isabstractmethod__": 3,
  "some_field": 2,
  "other_field": 2,
  "y": 2,
  "sins": 2,
  "__signature__": 2,
  "a": 1,
  "c": 1,
  "extra_field": 1,
  "_y": 1,
  "__wrapped__": 1,
  "_partialmethod": 1,
  "__code__": 1,
  "__defaults__": 1,
  "__kwdefaults__": 1
}
```
The most common one, `keys`, seems to be there because when you call `dict(model)` Python actually looks for an attribute called `keys` and uses it to build the dict if it exists. (Note that this causes a bizarre error converting an object implementing an appropriate `__iter__` into a dict if it has a non-callable attribute called `keys`; this isn't Pydantic-specific).

The others seem reasonable to me and most of them likely fall into one of the following buckets:
* purposeful attribute access of missing attributes in tests
* reading class attributes off instances
* checking for instances to implement interfaces via dunder methods

At any rate, none of the above seems likely to have performance implications, so I think we should be safe on that front.

**Review by hramezani — APPROVED** [review]

**Pull request merged**